app/services/arcgis_api/token_keeper.rb

```diff
@@ -0,0 +1,247 @@
+module ArcgisApi
+  # Struct to store token information, this allows us to track
+  # real expiration time with various rails cache backends many of them
+  # do not support entry expiration.
+  # @member token the token string
+  # @member expires_at hard expiration timestamp in epoch seconds
+  # @member sliding_expires_at optional the token keeper to maintain for
+  #   sliding expiration when sliding expiration enabled.
+  #   A time that the token does not actually expire
+  #   but used to control the timing of requesting a new token before it expires.
+  #   It's initially set to expires_at - 3*prefetch_ttl.
+  TokenInfo = Struct.new(
+    :token,
+    :expires_at,
+    :sliding_expires_at,
+  )
+
+  # Class for retrieve, refresh and manage caching of Arcgis API token.
+  # If token synchronous fetching is disabled, a token will be fetched from
+  # cache directly or it will be a cache miss.
+  # Otherwise the thread will try to fetch/refresh the token on demand as needed.
+  #
+  # When sliding_expires_at<= current time <= sliding_expires_at + prefetch_ttl,
+  # the entry's sliding_expires_at time is updated to
+  # sliding_expires_at + prefetch_ttl and a new token is requested and saved to cache.
+  #
+  # When sliding_expires_at + prefetch_ttl < current time,
+  # a new token is requested and saved to cache.
+  #
+  # Optimistically, the token in cache will NOT expire
+  # when accessed by multi-threaded/distributed clients,
+  # since there is about prefetch_ttl - 2*prefetch_ttl
+  # length of time to generate a new API token.
+
+  class TokenKeeper
+    API_TOKEN_HOST = URI(IdentityConfig.store.arcgis_api_generate_token_url).host
+    API_TOKEN_CACHE_KEY =
+      "#{IdentityConfig.store.arcgis_api_token_cache_key_prefix}:#{API_TOKEN_HOST}"
+    API_PREFETCH_TTL_SECONDS = IdentityConfig.store.arcgis_api_token_prefetch_ttl
+
+    RETRY_EXCEPTION = [404, 408, 409, 421, 429, 500, 502, 503, 504, 509]
```

Comment:
This is a little complicated to dig up, but based on my prior research we'll be dealing with errors here that are similar to [those mentioned in the ArcGIS REST JS documentation](https://developers.arcgis.com/arcgis-rest-js/error-handling/#generate-token-errors). You can view the underlying logic in [arcgis-rest-js/packages](https://github.com/Esri/arcgis-rest-js/blob/c950c4a9112c391e5f49204233bdbef9f65582d1/packages/arcgis-rest-request/src/request.ts#L148).

Mentioning this because I noticed a few additional errors here beyond what we're likely to see while using the ArcGIS API.